File: scripts/export_control_administrative.py
#!/usr/bin/env python3
"""
Export Control Administrative - Professional Excel Report
==========================================================
Uses xlsxwriter to generate a print-ready .xlsx with:
- add_table (filters + zebra striping)
- Dark blue header (#1f4e78) white bold centered text
- Percentage (0.00%%) for Failure Rate %%, Def Rate %%
- Thousands separator for Audits, Failures, Qty
- Conditional formatting: FAIL -> red, PASS -> green
- Page setup: landscape, fit to 1 page wide, 0.5" margins, header title
- Dynamic column widths

Usage:
  # From a JSON file exported by the system
  python scripts/export_control_administrative.py data.json -o reporte.xlsx

  # Pipe JSON via stdin
  python scripts/export_control_administrative.py -o reporte.xlsx < data.json

  # Use sample data (for testing)
  python scripts/export_control_administrative.py --sample -o reporte.xlsx
"""

import json, sys, argparse
import xlsxwriter


HEADERS = [
    "Fecha", "Fabrica", "Cliente", "Estilo", "PO",
    "Type", "Name", "Stage", "Point",
    "Audits", "Failures", "Failure Rate %",
    "Meas Insp", "Meas Def", "Meas Def Rate %",
    "Vis Insp", "Vis Def", "Vis Def Rate %",
    "Result",
]

# Index of numeric columns (0-based)
IDX_AUDITS = 9
IDX_FAILURES = 10
IDX_FAILURE_RATE = 11
IDX_MEAS_INSP = 12
IDX_MEAS_DEF = 13
IDX_MEAS_DEF_RATE = 14
IDX_VIS_INSP = 15
IDX_VIS_DEF = 16
IDX_VIS_DEF_RATE = 17
IDX_RESULT = 18

TOTAL_LABELS = {"TOTAL ", "Total ", "GRAN TOTAL", " Subtotal ", "  Subtotal "}


def is_subtotal_row(row):
    """Check if row is a subtotal/total marker; returns label type or None."""
    if not row or not row[0]:
        return None
    v = str(row[0])
    for label in TOTAL_LABELS:
        if v.startswith(label) or v == label:
            return label.strip()
    return None


def build_workbook(data_rows, output_path, title="CONTROL ADMINISTRATIVE - BASE DEPURADA"):
    """Create the xlsxwriter workbook with all formatting."""
    wb = xlsxwriter.Workbook(output_path)
    ws = wb.add_worksheet("Base Depurada")

    # Formats
    fmt_header = wb.add_format({
        "bold": True,
        "font_color": "white",
        "bg_color": "#1f4e78",
        "border": 1,
        "border_color": "#a0a0a0",
        "align": "center",
        "valign": "vcenter",
        "text_wrap": True,
        "font_name": "Calibri",
        "font_size": 10,
    })
    fmt_title = wb.add_format({
        "bold": True,
        "font_color": "#1f4e78",
        "bg_color": "#d6e4f0",
        "border": 1,
        "border_color": "#a0a0a0",
        "align": "center",
        "valign": "vcenter",
        "font_name": "Calibri",
        "font_size": 14,
    })
    fmt_pct = wb.add_format({
        "num_format": "0.00%",
        "border": 1,
        "border_color": "#a0a0a0",
        "align": "right",
        "valign": "vcenter",
        "font_name": "Calibri",
        "font_size": 10,
    })
    fmt_int = wb.add_format({
        "num_format": "#,##0",
        "border": 1,
        "border_color": "#a0a0a0",
        "align": "right",
        "valign": "vcenter",
        "font_name": "Calibri",
        "font_size": 10,
    })
    fmt_text = wb.add_format({
        "border": 1,
        "border_color": "#a0a0a0",
        "align": "left",
        "valign": "vcenter",
        "font_name": "Calibri",
        "font_size": 10,
    })
    fmt_text_center = wb.add_format({
        "border": 1,
        "border_color": "#a0a0a0",
        "align": "center",
        "valign": "vcenter",
        "font_name": "Calibri",
        "font_size": 10,
    })

    # Conditional: PASS (green)
    fmt_pass = wb.add_format({
        "bg_color": "#c6efce",
        "font_color": "#006100",
        "bold": True,
        "border": 1,
        "border_color": "#a0a0a0",
        "align": "center",
        "valign": "vcenter",
        "font_name": "Calibri",
        "font_size": 10,
    })
    # Conditional: FAIL (red)
    fmt_fail = wb.add_format({
        "bg_color": "#ffc7ce",
        "font_color": "#9c0006",
        "bold": True,
        "border": 1,
        "border_color": "#a0a0a0",
        "align": "center",
        "valign": "vcenter",
        "font_name": "Calibri",
        "font_size": 10,
    })

    # Alternating row formats (data section)
    fmt_data_even = wb.add_format({
        "bg_color": "#f0f4f8",
        "border": 1,
        "border_color": "#a0a0a0",
        "font_name": "Calibri",
        "font_size": 10,
    })
    fmt_data_odd = wb.add_format({
        "bg_color": "#ffffff",
        "border": 1,
        "border_color": "#a0a0a0",
        "font_name": "Calibri",
        "font_size": 10,
    })

    # Subtotal / total formats
    fmt_grand_total = wb.add_format({
        "bold": True,
        "font_color": "#ffffff",
        "bg_color": "#1e3a8a",
        "border": 1,
        "border_color": "#a0a0a0",
        "font_name": "Calibri",
        "font_size": 11,
    })
    fmt_month_total = wb.add_format({
        "bold": True,
        "font_color": "#dbeafe",
        "bg_color": "#1e40af",
        "border": 1,
        "border_color": "#a0a0a0",
        "font_name": "Calibri",
        "font_size": 10,
    })
    fmt_factory_total = wb.add_format({
        "bold": True,
        "font_color": "#d1fae5",
        "bg_color": "#065f46",
        "border": 1,
        "border_color": "#a0a0a0",
        "font_name": "Calibri",
        "font_size": 10,
    })
    fmt_buyer_subtotal = wb.add_format({
        "bold": True,
        "font_color": "#fef3c7",
        "bg_color": "#92400e",
        "border": 1,
        "border_color": "#a0a0a0",
        "font_name": "Calibri",
        "font_size": 10,
    })
    fmt_point_subtotal = wb.add_format({
        "bold": True,
        "font_color": "#ede9fe",
        "bg_color": "#5b21b6",
        "border": 1,
        "border_color": "#a0a0a0",
        "font_name": "Calibri",
        "font_size": 10,
    })

    # Write title row
    title_row = 0
    ws.merge_range(title_row, 0, title_row, len(HEADERS) - 1, title, fmt_title)
    ws.set_row(title_row, 30)

    # Column widths (dynamic estimate)
    col_widths = []
    for i, h in enumerate(HEADERS):
        max_w = len(h.encode("utf-8")) if h else 10
        for row in data_rows:
            val = str(row[i]) if i < len(row) and row[i] is not None else ""
            w = len(val.encode("utf-8"))
            if w > max_w:
                max_w = w
        clamped = max(min(max_w + 3, 40), 10)
        col_widths.append(clamped)
    for i, w in enumerate(col_widths):
        ws.set_column(i, i, w)

    # Write header row (row 1)
    header_row = 1
    for c, h in enumerate(HEADERS):
        ws.write(header_row, c, h, fmt_header)
    ws.set_row(header_row, 28)

    # Write data rows (starting row 2)
    data_start_row = header_row + 1
    last_data_row = data_start_row + len(data_rows) - 1

    for ri, row_data in enumerate(data_rows):
        excel_row = data_start_row + ri
        st = is_subtotal_row(row_data)

        # Pick format for the entire row based on subtotal type
        if st == "GRAN TOTAL":
            row_fmt_base = fmt_grand_total
            pct_fmt = fmt_grand_total
            int_fmt = fmt_grand_total
            text_fmt = fmt_grand_total
        elif st == "TOTAL" or (row_data[0] and str(row_data[0]).startswith("TOTAL ")):
            row_fmt_base = fmt_month_total
            pct_fmt = fmt_month_total
            int_fmt = fmt_month_total
            text_fmt = fmt_month_total
        elif row_data[0] and str(row_data[0]).startswith("Total "):
            row_fmt_base = fmt_factory_total
            pct_fmt = fmt_factory_total
            int_fmt = fmt_factory_total
            text_fmt = fmt_factory_total
        elif row_data[0] and " Subtotal " in str(row_data[0]):
            row_fmt_base = fmt_buyer_subtotal
            pct_fmt = fmt_buyer_subtotal
            int_fmt = fmt_buyer_subtotal
            text_fmt = fmt_buyer_subtotal
        elif row_data[0] and "  Subtotal " in str(row_data[0]):
            row_fmt_base = fmt_point_subtotal
            pct_fmt = fmt_point_subtotal
            int_fmt = fmt_point_subtotal
            text_fmt = fmt_point_subtotal
        else:
            is_even = ri % 2 == 0
            row_fmt_base = fmt_data_even if is_even else fmt_data_odd
            pct_fmt = row_fmt_base
            int_fmt = row_fmt_base
            text_fmt = row_fmt_base

        for c in range(len(HEADERS)):
            val = row_data[c] if c < len(row_data) else ""

            # Determine which format to use per cell
            if c == IDX_RESULT:
                if str(val).strip().upper() == "PASS":
                    ws.write(excel_row, c, val, fmt_pass)
                elif str(val).strip().upper() == "FAIL":
                    ws.write(excel_row, c, val, fmt_fail)
                else:
                    ws.write(excel_row, c, val, fmt_text_center)
            elif c in (IDX_FAILURE_RATE, IDX_MEAS_DEF_RATE, IDX_VIS_DEF_RATE):
                # Expect value as decimal (e.g. 0.10 for 10%)
                if isinstance(val, str) and val.endswith("%"):
                    val = float(val.replace("%", "")) / 100.0
                ws.write_number(excel_row, c, float(val or 0), pct_fmt)
            elif c in (IDX_AUDITS, IDX_FAILURES, IDX_MEAS_INSP, IDX_MEAS_DEF, IDX_VIS_INSP, IDX_VIS_DEF):
                ws.write_number(excel_row, c, int(val or 0), int_fmt)
            else:
                ws.write(excel_row, c, val, text_fmt)

    # add_table - creates proper Excel table with filters, banded rows, sortable headers
    table_range = "A{}:{}{}".format(header_row + 1, chr(65 + len(HEADERS) - 1), last_data_row + 1)
    ws.add_table(table_range, {
        "header_row": True,
        "banded_rows": True,
        "style": "Table Style Medium 9",
    })

    # Conditional formatting for Result column
    result_col_letter = chr(65 + IDX_RESULT) if IDX_RESULT < 26 else None
    if result_col_letter:
        result_range = "{}{}:{}{}".format(result_col_letter, data_start_row + 1, result_col_letter, last_data_row + 1)
        ws.conditional_format(result_range, {
            "type": "cell",
            "criteria": "equal to",
            "value": '"FAIL"',
            "format": fmt_fail,
        })
        ws.conditional_format(result_range, {
            "type": "cell",
            "criteria": "equal to",
            "value": '"PASS"',
            "format": fmt_pass,
        })

    # Page setup for printing
    ws.set_landscape()
    ws.set_paper(1)  # Letter
    ws.fit_to_pages(1, 0)  # 1 page wide, tall as needed
    ws.set_margins(left=0.5, right=0.5, top=0.5, bottom=0.5)
    ws.set_header("&C&B&14{}".format(title))
    ws.set_footer("&P / &N")
    ws.center_horizontally()
    ws.print_area(0, 0, last_data_row, len(HEADERS) - 1)

    wb.close()
    print("Report exported: " + output_path)


def generate_sample_data():
    """Generate sample data for testing."""
    import datetime, random

    rows = []
    factories = ["SAE-A TECHNOTEX, S.A(1)", "SAE-A TECHNOTEX, S.A(2)", "EINS, S.A.(2)"]
    buyers = ["NIKE", "ADIDAS", "PUMA", "UNDER ARMOUR"]
    stages = ["CUTTING", "SEWING", "PACKING"]
    points = ["IN-LINE", "FINAL", "RANDOM"]
    results = ["PASS", "FAIL"]

    for f in factories:
        for b in buyers:
            for s in stages:
                for p in points:
                    audits = random.randint(50, 500)
                    failures = random.randint(0, int(audits * 0.3))
                    meas_insp = random.randint(20, 200)
                    meas_def = random.randint(0, int(meas_insp * 0.2))
                    vis_insp = random.randint(20, 200)
                    vis_def = random.randint(0, int(vis_insp * 0.2))
                    failure_rate = failures / audits if audits > 0 else 0
                    meas_def_rate = meas_def / meas_insp if meas_insp > 0 else 0
                    vis_def_rate = vis_def / vis_insp if vis_insp > 0 else 0
                    result = "PASS" if random.random() > 0.3 else "FAIL"

                    rows.append({
                        "fecha": datetime.date(2025, random.randint(1, 12), random.randint(1, 28)).isoformat(),
                        "factory": f, "buyer": b, "style": "STY-" + str(random.randint(1000, 9999)),
                        "po": "PO-" + str(random.randint(10000, 99999)), "colL": "FABRIC", "colM": "",
                        "stage": s, "point": p,
                        "audits": audits, "failures": failures, "failureRate": failure_rate,
                        "measIns": meas_insp, "measDef": meas_def, "measDefRate": meas_def_rate,
                        "visIns": vis_insp, "visDef": vis_def, "visDefRate": vis_def_rate,
                        "result": result, "mesFiltroKey": "", "mesTexto": "",
                    })

    return rows


def data_to_rows(records):
    """Convert list of dicts (as from the system) to row arrays with subtotals."""
    rows = []
    L = len([
        "Fecha", "Fabrica", "Cliente", "Estilo", "PO",
        "Type", "Name", "Stage", "Point"
    ])

    def a0():
        return {"audits": 0, "failures": 0, "measIns": 0, "measDef": 0, "visIns": 0, "visDef": 0}

    def add_agg(t, s):
        t["audits"] += s["audits"]
        t["failures"] += s["failures"]
        t["measIns"] += s["measIns"]
        t["measDef"] += s["measDef"]
        t["visIns"] += s["visIns"]
        t["visDef"] += s["visDef"]

    def make_sub_row(agg, label):
        fr = agg["failures"] / agg["audits"] if agg["audits"] > 0 else 0
        mr = agg["measDef"] / agg["measIns"] if agg["measIns"] > 0 else 0
        vr = agg["visDef"] / agg["visIns"] if agg["visIns"] > 0 else 0
        return [label, "", "", "", "", "", "", "", "",
                agg["audits"], agg["failures"], fr,
                agg["measIns"], agg["measDef"], mr,
                agg["visIns"], agg["visDef"], vr, ""]

    def dr(r):
        fr = r["failures"] / r["audits"] if r["audits"] > 0 else 0
        mr = r["measDef"] / r["measIns"] if r["measIns"] > 0 else 0
        vr = r["visDef"] / r["visIns"] if r["visIns"] > 0 else 0
        return [r["fecha"], r["factory"], r["buyer"], r["style"], r["po"],
                r.get("colL", ""), r.get("colM", ""), r["stage"], r["point"],
                r["audits"], r["failures"], fr,
                r["measIns"], r["measDef"], mr,
                r["visIns"], r["visDef"], vr, r["result"]]

    # Group by month -> factory -> buyer -> point
    sorted_recs = sorted(records, key=lambda r: (
        r.get("mesFiltroKey", ""), r.get("factory", ""),
        r.get("buyer", ""), r.get("point", ""), r.get("fecha", "")
    ))

    pm = pf = pb = pp = ""
    am = af = ab = ap = a0()
    gt = a0()

    def fp():
        nonlocal pp, ap, ab
        if pp == "":
            return
        rows.append(make_sub_row(ap, "  Subtotal " + pp))
        add_agg(ab, ap)
        ap = a0()

    def fb():
        nonlocal pb, pp, ap, ab, af
        if pb == "":
            return
        fp()
        rows.append(make_sub_row(ab, " Subtotal " + pb))
        add_agg(af, ab)
        ab = a0()

    def ff():
        nonlocal pf, pb, pp, ap, ab, af, am
        if pf == "":
            return
        fb()
        rows.append(make_sub_row(af, "Total " + pf))
        add_agg(am, af)
        af = a0()

    def fm():
        nonlocal pm, pf, pb, pp, ap, ab, af, am, gt
        if pm == "":
            return
        ff()
        rows.append(make_sub_row(am, "TOTAL " + pm))
        add_agg(gt, am)
        am = a0()

    for r in sorted_recs:
        mk = r.get("mesFiltroKey", "")
        if mk != pm:
            fm()
            pm = mk
            pf = pb = pp = ""
        if r["factory"] != pf:
            ff()
            pf = r["factory"]
            pb = pp = ""
        if r["buyer"] != pb:
            fb()
            pb = r["buyer"]
            pp = ""
        if r["point"] != pp:
            fp()
            pp = r["point"]
        add_agg(ap, r)
        rows.append(dr(r))

    fp()
    fb()
    ff()
    fm()
    rows.append(make_sub_row(gt, "GRAN TOTAL"))

    # Remove empty spacer rows
    rows = [r for r in rows if r]

    return rows


def main():
    parser = argparse.ArgumentParser(
        description="Generate professional Control Administrative Excel report using xlsxwriter"
    )
    parser.add_argument("input", nargs="?", help="JSON input file (omit for stdin)")
    parser.add_argument("-o", "--output", default="control_administrative.xlsx",
                        help="Output .xlsx path")
    parser.add_argument("--sample", action="store_true",
                        help="Use sample data for testing")
    args = parser.parse_args()

    if args.sample:
        records = generate_sample_data()
        print("[INFO] Using sample data")
    elif args.input:
        with open(args.input, "r", encoding="utf-8") as f:
            data = json.load(f)
        records = data if isinstance(data, list) else data.get("entries", data.get("data", []))
        print("[INFO] Loaded {} records from {}".format(len(records), args.input))
    else:
        data = json.load(sys.stdin)
        records = data if isinstance(data, list) else data.get("entries", data.get("data", []))
        print("[INFO] Loaded {} records from stdin".format(len(records)))

    rows = data_to_rows(records)
    print("[INFO] Generated {} rows (including subtotals)".format(len(rows)))
    build_workbook(rows, args.output)


if __name__ == "__main__":
    main()

File: lib/export-control-administrative.ts
import ExcelJS from 'exceljs';

// ─── Types ──────────────────────────────────────────────────────────

export interface CARecord {
  fecha?: string;
  factory?: string;
  buyer?: string;
  style?: string;
  po?: string;
  colL?: string;
  colM?: string;
  stage?: string;
  point?: string;
  audits?: number;
  failures?: number;
  failureRate?: number;
  measIns?: number;
  measDef?: number;
  measDefRate?: number;
  visIns?: number;
  visDef?: number;
  visDefRate?: number;
  result?: string;
  mesFiltroKey?: string;
  mesTexto?: string;
  [key: string]: unknown;
}

const HEADERS = [
  'Fecha', 'Fabrica', 'Cliente', 'Estilo', 'PO',
  'Type', 'Name', 'Stage', 'Point',
  'Audits', 'Failures', 'Failure Rate %',
  'Meas Insp', 'Meas Def', 'Meas Def Rate %',
  'Vis Insp', 'Vis Def', 'Vis Def Rate %',
  'Result',
];

const IDX = {
  AUDITS: 9,
  FAILURES: 10,
  FAILURE_RATE: 11,
  MEAS_INSP: 12,
  MEAS_DEF: 13,
  MEAS_DEF_RATE: 14,
  VIS_INSP: 15,
  VIS_DEF: 16,
  VIS_DEF_RATE: 17,
  RESULT: 18,
};

interface Agg {
  audits: number;
  failures: number;
  measIns: number;
  measDef: number;
  visIns: number;
  visDef: number;
}

function zeroAgg(): Agg {
  return { audits: 0, failures: 0, measIns: 0, measDef: 0, visIns: 0, visDef: 0 };
}

function addAgg(t: Agg, s: Agg): void {
  t.audits += s.audits;
  t.failures += s.failures;
  t.measIns += s.measIns;
  t.measDef += s.measDef;
  t.visIns += s.visIns;
  t.visDef += s.visDef;
}

function subRow(agg: Agg, label: string): (string | number)[] {
  const fr = agg.audits > 0 ? agg.failures / agg.audits : 0;
  const mr = agg.measIns > 0 ? agg.measDef / agg.measIns : 0;
  const vr = agg.visIns > 0 ? agg.visDef / agg.visIns : 0;
  return [
    label, '', '', '', '', '', '', '', '',
    agg.audits, agg.failures, fr,
    agg.measIns, agg.measDef, mr,
    agg.visIns, agg.visDef, vr, '',
  ];
}

function dataRow(r: CARecord): (string | number)[] {
  const fr = r.audits! > 0 ? r.failures! / r.audits! : 0;
  const mr = r.measIns! > 0 ? r.measDef! / r.measIns! : 0;
  const vr = r.visIns! > 0 ? r.visDef! / r.visIns! : 0;
  return [
    r.fecha ?? '', r.factory ?? '', r.buyer ?? '', r.style ?? '', r.po ?? '',
    r.colL ?? '', r.colM ?? '', r.stage ?? '', r.point ?? '',
    r.audits ?? 0, r.failures ?? 0, fr,
    r.measIns ?? 0, r.measDef ?? 0, mr,
    r.visIns ?? 0, r.visDef ?? 0, vr, r.result ?? '',
  ];
}

function buildRows(records: CARecord[]): (string | number)[][] {
  const rows: (string | number)[][] = [];

  // Group by monthKey → factory → buyer → point
  const sorted = [...records].sort((a, b) => {
    const mk = (a.mesFiltroKey ?? '').localeCompare(b.mesFiltroKey ?? '');
    if (mk !== 0) return mk;
    const fk = (a.factory ?? '').localeCompare(b.factory ?? '');
    if (fk !== 0) return fk;
    const bk = (a.buyer ?? '').localeCompare(b.buyer ?? '');
    if (bk !== 0) return bk;
    const pk = (a.point ?? '').localeCompare(b.point ?? '');
    if (pk !== 0) return pk;
    return (a.fecha ?? '').localeCompare(b.fecha ?? '');
  });

  let pm = '', pf = '', pb = '', pp = '';
  let am = zeroAgg(), af = zeroAgg(), ab = zeroAgg(), ap = zeroAgg();
  const gt = zeroAgg();

  function flushPoint() {
    if (pp === '') return;
    rows.push(subRow(ap, `  Subtotal ${pp}`));
    addAgg(ab, ap);
    ap = zeroAgg();
  }

  function flushBuyer() {
    if (pb === '') return;
    flushPoint();
    rows.push(subRow(ab, ` Subtotal ${pb}`));
    addAgg(af, ab);
    ab = zeroAgg();
  }

  function flushFactory() {
    if (pf === '') return;
    flushBuyer();
    rows.push(subRow(af, `Total ${pf}`));
    addAgg(am, af);
    af = zeroAgg();
  }

  function flushMonth() {
    if (pm === '') return;
    flushFactory();
    rows.push(subRow(am, `TOTAL ${pm}`));
    addAgg(gt, am);
    am = zeroAgg();
  }

  for (const r of sorted) {
    const mk = r.mesFiltroKey ?? '';
    if (mk !== pm) { flushMonth(); pm = mk; pf = pb = pp = ''; }
    if (r.factory !== pf) { flushFactory(); pf = r.factory ?? ''; pb = pp = ''; }
    if (r.buyer !== pb) { flushBuyer(); pb = r.buyer ?? ''; pp = ''; }
    if (r.point !== pp) { flushPoint(); pp = r.point ?? ''; }
    addAgg(ap, r as Agg);
    rows.push(dataRow(r));
  }

  flushPoint();
  flushBuyer();
  flushFactory();
  flushMonth();
  rows.push(subRow(gt, 'GRAN TOTAL'));

  return rows.filter(r => r.length > 0);
}

function isSubtotalRow(row: (string | number)[]): string | null {
  const v = String(row[0] ?? '');
  if (v === 'GRAN TOTAL') return 'GRAN TOTAL';
  if (v.startsWith('TOTAL ')) return 'TOTAL';
  if (v.startsWith('Total ')) return 'Total';
  if (v.includes(' Subtotal ')) return ' Subtotal ';
  if (v.includes('  Subtotal ')) return '  Subtotal ';
  return null;
}

// ─── XLSX Generator (Professional) ──────────────────────────────────

export async function generateControlAdministrativeXLSX(
  records: CARecord[],
  title = 'CONTROL ADMINISTRATIVE - BASE DEPURADA',
): Promise<Uint8Array> {
  const rows = buildRows(records);
  const wb = new ExcelJS.Workbook();
  wb.creator = 'System JB - Control Administrative';
  wb.created = new Date();
  const ws = wb.addWorksheet('Base Depurada');

  // Column widths
  const colWidths = HEADERS.map((h, i) => {
    let maxW = h.length;
    for (const r of rows) {
      const val = String(r[i] ?? '');
      maxW = Math.max(maxW, val.length);
    }
    return Math.max(10, Math.min(maxW + 3, 40));
  });
  for (let i = 0; i < colWidths.length; i++) ws.getColumn(i + 1).width = colWidths[i];

  let row = 1;

  // Title row
  ws.mergeCells(row, 1, row, HEADERS.length);
  const titleCell = ws.getCell(row, 1);
  titleCell.value = title;
  titleCell.font = { bold: true, color: { argb: 'FF1F4E78' }, name: 'Calibri', size: 14 };
  titleCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFD6E4F0' } };
  titleCell.alignment = { horizontal: 'center', vertical: 'middle' };
  titleCell.border = { top: { style: 'thin', color: { argb: 'FFA0A0A0' } }, bottom: { style: 'thin', color: { argb: 'FFA0A0A0' } }, left: { style: 'thin', color: { argb: 'FFA0A0A0' } }, right: { style: 'thin', color: { argb: 'FFA0A0A0' } } };
  ws.getRow(row).height = 30;
  row++;

  // Header row
  for (let c = 0; c < HEADERS.length; c++) {
    const cell = ws.getCell(row, c + 1);
    cell.value = HEADERS[c];
    cell.font = { bold: true, color: { argb: 'FFFFFFFF' }, name: 'Calibri', size: 10 };
    cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF1F4E78' } };
    cell.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
    cell.border = { top: { style: 'thin', color: { argb: 'FFA0A0A0' } }, bottom: { style: 'thin', color: { argb: 'FFA0A0A0' } }, left: { style: 'thin', color: { argb: 'FFA0A0A0' } }, right: { style: 'thin', color: { argb: 'FFA0A0A0' } } };
  }
  ws.getRow(row).height = 28;
  row++;

  // Data rows
  const dataStartRow = row;
  for (let ri = 0; ri < rows.length; ri++) {
    const rowData = rows[ri];
    const st = isSubtotalRow(rowData);

    const getColors = () => {
      if (st === 'GRAN TOTAL') return { bg: 'FF1E3A8A', fg: 'FFFFFFFF', bold: true };
      if (st === 'TOTAL') return { bg: 'FF1E40AF', fg: 'FFDBEAFE', bold: true };
      if (st === 'Total') return { bg: 'FF065F46', fg: 'FFD1FAE5', bold: true };
      if (st === ' Subtotal ') return { bg: 'FF92400E', fg: 'FFFEF3C7', bold: true };
      if (st === '  Subtotal ') return { bg: 'FF5B21B6', fg: 'FFEDE9FE', bold: true };
      return { bg: ri % 2 === 0 ? 'FFF0F4F8' : 'FFFFFFFF', fg: 'FF000000', bold: false };
    };

    for (let c = 0; c < HEADERS.length; c++) {
      const cell = ws.getCell(row, c + 1);
      const val = rowData[c] ?? '';
      const colors = getColors();
      const isPct = c === IDX.FAILURE_RATE || c === IDX.MEAS_DEF_RATE || c === IDX.VIS_DEF_RATE;
      const isInt = c === IDX.AUDITS || c === IDX.FAILURES || c === IDX.MEAS_INSP || c === IDX.MEAS_DEF || c === IDX.VIS_INSP || c === IDX.VIS_DEF;
      const isResult = c === IDX.RESULT;
      const sv = String(val).toUpperCase().trim();
      const isPass = isResult && sv === 'PASS';
      const isFail = isResult && sv === 'FAIL';

      cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: isPass ? 'FFC6EFCE' : isFail ? 'FFFFC7CE' : colors.bg } };
      cell.font = {
        color: { argb: isPass ? 'FF006100' : isFail ? 'FF9C0006' : colors.fg },
        name: 'Calibri',
        size: colors.bold ? (st === 'GRAN TOTAL' ? 11 : 10) : 10,
        bold: isPass || isFail || colors.bold,
      };
      cell.alignment = {
        horizontal: isResult || (isPct && !colors.bold) ? 'center' : isPct || isInt ? 'right' : 'left',
        vertical: 'middle',
      };
      cell.border = { top: { style: 'thin', color: { argb: 'FFA0A0A0' } }, bottom: { style: 'thin', color: { argb: 'FFA0A0A0' } }, left: { style: 'thin', color: { argb: 'FFA0A0A0' } }, right: { style: 'thin', color: { argb: 'FFA0A0A0' } } };

      if (isPct) {
        cell.value = Number(val);
        cell.numFmt = '0.00%';
      } else if (isInt) {
        cell.value = Number(val);
        cell.numFmt = '#,##0';
      } else {
        cell.value = val;
      }
    }
    row++;
  }

  // Page setup
  ws.pageSetup.orientation = 'landscape';
  ws.pageSetup.paperSize = 9; // Letter
  ws.pageSetup.fitToPage = true;
  ws.pageSetup.fitToWidth = 1;
  ws.pageSetup.margins = { left: 0.5, right: 0.5, top: 0.5, bottom: 0.5, header: 0, footer: 0 };

  const buf = await wb.xlsx.writeBuffer();
  return buf instanceof Blob ? new Uint8Array(await buf.arrayBuffer()) : new Uint8Array(buf);
}
